File: day02/q1.py
def solve(input_data: str) -> str:
	nums = set()
	for row in input_data.split(","):
		start, end = tuple([ int(val) for val in row.split("-") ])
		for num in range(start, end + 1):
			nums.add(num) if symmetrical(num) else None
	return sum(nums)

def symmetrical(num: int):
	s = str(num)
	length = len(s)
	halflength = int(length / 2)
	return length % 2 == 0 and s[0:halflength] == s[halflength:length]

File: day06/q1.ts
export const solve = (inputData: string) => {
	const lines = inputData.split('\n').filter(line => line.length > 0);
	const colSpec = getColPositions(lines.splice(lines.length - 1, 1)[0]!)
	return colSpec.reduce((sum, { startIdx, endIdxExcl, op }) => {
		return sum + (
			op === '+'
				? lines.reduce((acc, row, idx) => idx ? Number(row.slice(startIdx, endIdxExcl)) + acc : Number(row.slice(startIdx, endIdxExcl)), 0)
				: lines.reduce((acc, row, idx) => idx ? Number(row.slice(startIdx, endIdxExcl)) * acc : Number(row.slice(startIdx, endIdxExcl)), 0)
		)
	}, 0)
}

const getColPositions = (lastLine: string) => {
	return lastLine
		.split('')
		.reduce((opPosIndexes, cur, curIndex, line) => {
			if (cur !== ' ') {
				if (opPosIndexes.length) {
					opPosIndexes[opPosIndexes.length - 1]!.endIdxExcl = curIndex
				}
				opPosIndexes.push({ op: cur as '*' | '+', startIdx: curIndex, endIdxExcl: Infinity })
			}
			return opPosIndexes
		}, <{startIdx: number, op: '+' | '*', endIdxExcl: number}[]>[])
}